### PR TITLE
Rate function added 

### PR DESCRIPTION
**Pull request by thsubaku9**
This PR adds the Rate calculation function to go-financial (using Newton Rapson for the root calculation)

### CODE REVIEW

**Comment by thsubaku9** [comment]
@gyanesh-m sorry to bother you but I think something is wrong with the CI. Do let me know when things are fixed so that I can continue with the PR

File: reducing_utils.go [hunk 1]
```diff
@@ -290,3 +290,83 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+ nper 	: number of compounding periods
+ pmt	: a (fixed) payment, paid either
+	  	  at the beginning (when = 1) or the end (when = 0) of each period
+ pv		: a present value
+ fv		: a future value
+ when 	: specification of whether payment is made
+		  at the beginning (when = 1) or the end (when = 0) of each period
+ curRate: the rate compounded once per period rate
+*/
+func getRateRatio(pv, fv, pmt, curRate float64, nper int64, when paymentperiod.Type) float64 {
```
Comment (gyanesh-m) [inline reducing_utils.go:307]: don't use float, use decimal. Check other functions in this file.

File: reducing_utils.go [hunk 2]
```diff
@@ -290,3 +290,83 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+	Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May). Open Document
+    Format for Office Applications (OpenDocument)v1.2, Part 2: Recalculated
+    Formula (OpenFormula) Format - Annotated Version, Pre-Draft 12.
+    Organization for the Advancement of Structured Information Standards
+    (OASIS). Billerica, MA, USA. [ODT Document]. Available:
+    http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formula
+    OpenDocument-formula-20090508.odt
+*/
+
+func Rate(pv, fv, pmt float64, nper int64, when paymentperiod.Type, params ...float64) (float64, bool) {
```
Comment (gyanesh-m) [inline reducing_utils.go:342]: same, replace float with decimal.

**Comment by gyanesh-m** [comment]
@thsubaku9 The pipeline is fine. There are errors in golangci-lint check. Run them locally using `make lint-check`. Make sure you have golangci-lint installed. If not, use `make install-lint` to install it.

File: reducing_utils.go (hunk 2 above)
Comment (thsubaku9) [reply reducing_utils.go]: oops sorry about that. But also glad to see this library now supports floating precession :)

**Comment by thsubaku9** [comment]
Let me know if any more changes are required @gyanesh-m

**Comment by gyanesh-m** [comment]
@thsubaku9 will review the changes over weekend.

File: reducing_utils.go [hunk 3]
```diff
@@ -290,3 +290,97 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+	Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May). Open Document
+    Format for Office Applications (OpenDocument)v1.2, Part 2: Recalculated
+    Formula (OpenFormula) Format - Annotated Version, Pre-Draft 12.
+    Organization for the Advancement of Structured Information Standards
+    (OASIS). Billerica, MA, USA. [ODT Document]. Available:
+    http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formula
+    OpenDocument-formula-20090508.odt
+*/
+
+func Rate(pv, fv, pmt decimal.Decimal, nper int64, when paymentperiod.Type, params ...decimal.Decimal) (decimal.Decimal, bool) {
```
Comment (gyanesh-m) [inline reducing_utils.go:356]: keep maxIter, initialGuess and tolerance as regular params and not variadic.

Comment (thsubaku9) [reply reducing_utils.go]: Sorry about that. Didn't know whether to keep it strict or variadic. Changes have been reflected in code, readme and play ground link as well

**Comment by thsubaku9** [comment]
lets gooooooo @gyanesh-m

File: README.md [hunk 4]
```diff
@@ -573,3 +575,66 @@ func main() {
 }
 ```
 [Run on go-playground](https://play.golang.org/p/hm77MTPBGYg)
+
+## Rate  
+
+```go  
+func Rate(pv, fv, pmt decimal, nper int64, when paymentperiod.Type, params ...float64) (decimal, bool)
```
Comment (gyanesh-m) [inline README.md:582]: update.

File: reducing_utils_test.go [hunk 5]
```diff
@@ -355,3 +355,54 @@ func Test_Nper(t *testing.T) {
...
+				pmt:          decimal.NewFromInt(100),
+				nper:         4,
+				when:         paymentperiod.BEGINNING,
+				maxIter:      100,
+				tolerance:    decimal.NewFromFloat(1e-7),
+				initialGuess: decimal.NewFromFloat(0.1),
+			},
+			want: decimal.NewFromFloat(0.06106257989825202),
+		}, {
+			name: "success", args: args{
```
Comment (gyanesh-m) [inline reducing_utils_test.go:388]: add nan test case.

Comment (thsubaku9) [reply reducing_utils_test.go:390]: so nan basically requires your validity to be false, so I've changed the test to handle that case as well

**Comment by thsubaku9** [comment]
Lets go ? @gyanesh-m

File: reducing_utils.go [hunk 6]
```diff
@@ -290,3 +290,82 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+ tolerance : accept result only if the difference in iteration values is less than the tolerance provided
+ initialGuess : an initial point to start approximating from
+References:
+	Wheeler, D. A., E. Rathke, and R. Weir (Eds.) (2009, May). Open Document
+    Format for Office Applications (OpenDocument)v1.2, Part 2: Recalculated
+    Formula (OpenFormula) Format - Annotated Version, Pre-Draft 12.
+    Organization for the Advancement of Structured Information Standards
+    (OASIS). Billerica, MA, USA. [ODT Document]. Available:
+    http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formula
+    OpenDocument-formula-20090508.odt
```
Comment (gyanesh-m) [inline reducing_utils.go:355]: fix indentation around **:**
Run `godoc -http=:80`
and go to http://localhost/pkg/github.com/razorpay/go-financial/#Rate to view the mismatched 
 indentations.

File: reducing_utils.go [hunk 7]
```diff
@@ -290,3 +290,82 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+*/
+
```
Comment (gyanesh-m) [inline reducing_utils.go:357]: no newline before a function comment. Otherwise it gets skipped.
```suggestion
```

File: reducing_utils.go [hunk 8]
```diff
@@ -290,3 +290,82 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+func Rate(pv, fv, pmt decimal.Decimal, nper int64, when paymentperiod.Type, maxIter int64, tolerance, initialGuess decimal.Decimal) (decimal.Decimal, bool) {
```
Comment (gyanesh-m) [inline reducing_utils.go:358]:
```suggestion
func Rate(pv, fv, pmt decimal.Decimal, nper int64, when paymentperiod.Type, maxIter int64, tolerance, initialGuess decimal.Decimal) (decimal.Decimal, error) {
```
also add error code for nan output in `go-financial/error_codes.go`

File: reducing_utils.go [hunk 9]
```diff
@@ -290,3 +290,82 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+	var nextIterRate, currentIterRate decimal.Decimal = initialGuess, initialGuess
+
+	for iter := int64(0); iter < maxIter; iter++ {
+		currentIterRate = nextIterRate
+		nextIterRate = currentIterRate.Sub(getRateRatio(pv, fv, pmt, currentIterRate, nper, when))
+	}
+
+	if nextIterRate.Sub(currentIterRate).Abs().GreaterThan(tolerance) {
+		return nextIterRate, false
```
Comment (gyanesh-m) [inline reducing_utils.go:367]: return decimal.Zero with custom error code.

File: reducing_utils.go [hunk 10]
```diff
@@ -290,3 +290,82 @@ func Npv(rate decimal.Decimal, values []decimal.Decimal) decimal.Decimal {
...
+    Organization for the Advancement of Structured Information Standards
+    (OASIS). Billerica, MA, USA. [ODT Document]. Available:
+    http://www.oasis-open.org/committees/documents.php?wg_abbrev=office-formula
+    OpenDocument-formula-20090508.odt
+*/
+
+func Rate(pv, fv, pmt decimal.Decimal, nper int64, when paymentperiod.Type, maxIter int64, tolerance, initialGuess decimal.Decimal) (decimal.Decimal, bool) {
+	var nextIterRate, currentIterRate decimal.Decimal = initialGuess, initialGuess
+
+	for iter := int64(0); iter < maxIter; iter++ {
```
Comment (gyanesh-m) [inline reducing_utils.go:361]: add tolerance check(https://github.com/numpy/numpy-financial/blob/master/numpy_financial/_financial.py#L661) along with iter<maxIter condition.
Refer: https://github.com/numpy/numpy-financial/blob/master/numpy_financial/_financial.py#L661

File: reducing_utils.go (hunk 9 above)
Comment (gyanesh-m) [reply reducing_utils.go:367]: also, make it `GreaterThanOrEqual`

File: reducing_utils.go (hunk 8 above)
Comment (thsubaku9) [reply reducing_utils.go]: Alright, errors are a better way to go about it

File: README.md [hunk 11]
```diff
@@ -627,14 +627,14 @@ func main() {
 	tolerance := decimal.NewFromFloat(1e-6)
 	initialGuess := decimal.NewFromFloat(0.1),
 
-	rate, ok := gofinancial.Rate(pv, fv, pmt, nper, when, maxIter, tolerance, initialGuess)
-	if ok {
-		fmt.Printf("rate:%v ", rate)
-	} else {
+	rate, err := gofinancial.Rate(pv, fv, pmt, nper, when, maxIter, tolerance, initialGuess)
+	if err != nil {
 		fmt.Printf("NaN")
```
Comment (gyanesh-m) [inline README.md:632]: print error.

**Comment by thsubaku9** [comment]
Any updates here ? @krantideep95 @gyanesh-m  @ra0van

**Comment by gyanesh-m** [comment]
Hey @thsubaku9, I will review this over the coming weekend.

**Review by gyanesh-m — APPROVED** [review]
lgtm

**Comment by gyanesh-m** [comment]
@thsubaku9 Thank you for adding this to go-financial ! 💯  🎉

**Pull request merged**